### ISSUE TITLE
Allow Custom Data Attachments in AssetInfo

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#285]
Right now, the [AssetInfo](https://github.com/AdamsLair/duality/blob/master/Duality/Resources/AssetInfo.cs) class stores a hardcoded set of information. However, it would be very nice if AssetImporters could store additional information in there:

- The original file format this Resource was imported from, so it can be exported to a matching file.
- Custom import parameters, obtained by user interaction and re-used when available.
- It could actually replace the current `Font.EmbeddedTrueTypeFont` property, which would be a cleaner design anyway.
- ...others?

**Comment by ilexp** [comment]
Design idea:

- Adjust the ResourcePropertyEditor to detect if there is additional data attached to the AssetInfo object and display a specialized AssetInfoPropertyEditor for this after the one for the actual resource.
  - This would allow to edit the settings that are used by the importer for re-import operations.
  - There should be a ReImport button at the end of this editor.
  - There should be a way to specify which items are displayed from within the importer.
  - Maybe the property editor could ask the preferred importer (as already specified in AssetInfo) for a settings object to display.
- This could actually solve multiple problems at once and make a very convenient way to configure Asset re-import operations after they've been imported for the first time.

**Comment by ilexp** [comment]
**Progress:**

- Did a prototype implementation and discarded it again.

**Issue:**

- When defining editor-specific `AssetImporterConfig` classes that will be serialized as part of a `Resource`, there will be serialization errors when loading that `Resource` with the required editor plugin removed in a release.
- This will also restrict resources to only being loaded only after all editor plugins have been loaded as well.
- None of the above is acceptable.

**Solution:**

- Instead of letting an editor plugin define a class that holds an importer's configuration, split up the configuration of an import into N key-value pairs mapping from string (id) to object (value).
  - No knowledge about editor classes is required for serialization.
- The import / export environment can provide specialized API for accessing and modifying this configuration.
  - The API should handle all the casting and availability checking conveniently.
  - This API could also check newly assigned object values for their type to make sure that no editor plugin type is used, prompting the above issue on the user side.
- An `AssetInfoPropertyEditor` implementation would have to manually populate its property editors based on the currently available key-value pairs.
  - Dynamically adding and removing those pairs is not required, as it would not have been a requirement with config classes.
  - Editor visibility of individual pairs can be done using the naming convention that was introduced for shader variable hiding: Adding an underscore in the beginning flags a value as "private".

As an additional advantage, different importers could share configuration data without having a hard reference to each others editor plugin. This would be useful when switching from one importer to another without having to reconfigure all resources.

**Comment by mika76** [comment]
Heh, now that I've hit the in-editor Reload Code you speak about I can totally relate to this being a pain. I wonder, what other use-cases are there other than for embedded fonts?

**Comment by ilexp** [comment]
> Heh, now that I've hit the in-editor Reload Code you speak about I can totally relate to this being a pain. I wonder, what other use-cases are there other than for embedded fonts?

Any kind of import config really. 

I don't have specific other use cases in mind, except for the embedded font thing, but I'd say complex or ambiguous source data has a good chance of making use of this. Say, for example, someone writes an importer that takes a Tiled map as a source file and generates and / or a Duality Scene from it: I'm pretty sure there is lots of optional stuff that could be included or not, and other stuff that might require additional information.

**Comment by ilexp** [comment]
**Notes:**

- The Font import could, once this system is implemented, actually do without any kind of embedded truetype font information.
  - Re-Import operations can only be done when the `Source/Media` files required for it are still there.
  - If the source ttf is still there, no need to embed it.
  - All the font importer would store in the data attachment would be the settings on how to import it, such as size, bold / italic, etc.
- For a **proper Re-Import setup**, there needs to be a way to determine which files to use as a base for the re-import operation.
  - Could do a simulated export to determine file paths. This requires symmetrical import / export, which isn't always available.
  - There should be a way to determine source file paths of a Resource via the AssetImporter.
    - Combination of AssetManagement API and data attachment hints?
- There should be **better support for importers without an export function**. It's a common case.
  - Don't rely on simulated export that much.
  - Introduce a distinct way to determine which `Source/Media` files belong to a given Resource.
    - As mentioned above, combination of AssetManagement API and data attachment hints could do the trick.
    - Look into code a little more to figure out the details.

**Comment by mika76** [comment]
Could you also do another thing? Please add a way to add custom data in the importer prepare step, which propagates to the import set (and probably vice versa in the export step), maybe against each input even.

In my case, I would like to parse the xml BMFont file (for example) read some stuff from it and validate it, and then if all is well I would add the asset (ImportPrepare) - but in the actual Import step I now need to re-read the file to get the info.

**Comment by ilexp** [comment]
> Could you also do another thing? Please add a way to add custom data in the importer prepare step, which propagates to the import set (and probably vice versa in the export step), maybe against each input even.

> In my case, I would like to parse the xml BMFont file (for example) read some stuff from it and validate it, and then if all is well I would add the asset (ImportPrepare) - but in the actual Import step I now need to re-read the file to get the info.

Sounds reasonable, but I think it should be distinct from the custom data attachments in this issue, since they are (a) per-Resource and (b) meant to be serialized with the Resource.

Also, the prepare step is meant to skim over the available input data, decide whether the importer can handle and of this and then predict which files the importer will use to create which Resources as an output. I'm not sure whether any validation should be performed here. A BMFont importer is responsible for a BMFont regardless of whether or not that BMFont is actually valid, so the preparation step should perform all the same. If you encounter an invalid font, it's the import step itself that should fail.

Think of it like this: If you have an invalid `.png` file on your drive, your paint software will display it in the "Open File" dialog (the prepare step) but then fail to actually open it (the import step), which is what you'd expect. 

With the above in mind, I'd rather postpone potential additions regarding per-operation temporary importer data API. Maybe there isn't actually a use case for them.

**Comment by mika76** [comment]
The main reason I perform the parsing of the xml file is because the name of the actual font pixel file is in it. And the xml files (having so many different formats) all have the same extensions either fnt or xml so it's kind of hard to deduce anything unless the file is actually read. And to read I have to figure out the type hence the validation. :smile: (it's been fun)

I did mean totally temporary / non serializable memory purely for access data between steps.

Of course the one question is can I replace an import object with my own, inherited one? In which case nothing new is needed...

**Comment by ilexp** [comment]
> The main reason I perform the parsing of the xml file is because the name of the actual font pixel file is in it. And the xml files (having so many different formats) all have the same extensions either fnt or xml so it's kind of hard to deduce anything unless the file is actually read. And to read I have to figure out the type hence the validation. :smile: (it's been fun)

Ah, that makes sense. I'll file a new issue for this. Am I correctly assuming that this is a non-critical performance improvement from your perspective?

> Of course the one question is can I replace an import object with my own, inherited one? In which case nothing new is needed.

You could, but that wouldn't mesh well with the editor-wide AssetManagement system, so better don't do this.

**Comment by mika76** [comment]
> Am I correctly assuming that this is a non-critical performance improvement from your perspective?

Yup - currently I just re-read and re-validate.

**Comment by ilexp** [comment]
Created issue #318 that will deal distinctly with passing data from prepare to import. Also created issue #319 that will deal with introducing better support for non-symmetrical import / export.

**Comment by mika76** [comment]
Cool thanks :smile:

**Comment by ilexp** [comment]
Added custom data attachments to `AssetInfo` and provided a new parameter retrieval API for asset import / export environments that uses the custom data attachment as a backend.

Closing this.
